Support device as Dict key

pbelevich opened the pull request:
This is required to use RPC DeviceMap aka Dict[torch.device, torch.device] in torchscript

Stack from [ghstack](https://github.com/ezyang/ghstack):
* #64901
* #65164
* __->__ #65079

Differential Revision: [D31072626](https://our.internmc.facebook.com/intern/diff/D31072626)

pbelevich committed:
[JIT] Support device as Dict key

[ghstack-poisoned]

pbelevich committed:
Update on "[JIT] Support device as Dict key"

This is required to use RPC DeviceMap aka Dict[torch.device, torch.device] in torchscript




[ghstack-poisoned]

pbelevich committed:
Update on "[JIT] Support device as Dict key"

This is required to use RPC DeviceMap aka Dict[torch.device, torch.device] in torchscript




[ghstack-poisoned]

pbelevich committed:
Update on "[JIT] Support device as Dict key"

This is required to use RPC DeviceMap aka Dict[torch.device, torch.device] in torchscript




[ghstack-poisoned]

pbelevich committed:
Update on "[JIT] Support device as Dict key"

This is required to use RPC DeviceMap aka Dict[torch.device, torch.device] in torchscript




[ghstack-poisoned]

pbelevich committed:
Update on "[JIT] Support device as Dict key"

This is required to use RPC DeviceMap aka Dict[torch.device, torch.device] in torchscript




[ghstack-poisoned]

pbelevich committed:
Update on "[JIT] Support device as Dict key"

This is required to use RPC DeviceMap aka Dict[torch.device, torch.device] in torchscript




[ghstack-poisoned]

lw commented on aten/src/ATen/core/Dict_inl.h:
```diff
@@ -50,6 +50,8 @@ inline size_t DictKeyHash::operator()(const IValue& ivalue) const {
   } else if (ivalue.isTensor()) {
     return std::hash<TensorImpl*>()(ivalue.toTensor().unsafeGetTensorImpl());
+  } else if (ivalue.isDevice()) {
+    return std::hash<std::string>()(ivalue.toDevice().str());
```
There already exists a specialization of `std::hash` for devices: https://github.com/pytorch/pytorch/blob/b9fc656cf26d60127bd695e4e5a7d27622f2563d/c10/core/Device.h#L137

Hence this could be simplified, and made more idiomatic and efficient, by doing:
```suggestion
    return std::hash<Device>()(ivalue.toDevice());
```

pbelevich replied:
Thanks for suggestion!

lw reviewed (approved):
Overall looks good to me (with one suggestion).

pbelevich committed:
Update on "[JIT] Support device as Dict key"

This is required to use RPC DeviceMap aka Dict[torch.device, torch.device] in torchscript




[ghstack-poisoned]

pbelevich commented:
@pbelevich has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31072626).
